include/simdjson/numberparsing.h

```diff
@@ -536,34 +500,40 @@ static really_inline bool parse_number(const uint8_t *const buf, ParsedJson &pj,
       exp_number = 10 * exp_number + digit;
       ++p;
     }
-    exponent += (neg_exp ? -exp_number : exp_number);
+    exp_number = (neg_exp ? -exp_number : exp_number);
   }
   if (is_float) {
-    uint64_t power_index = 308 + exponent;
-    if (unlikely((digit_count >= 19))) { // this is uncommon
-      // It is possible that the integer had an overflow.
-      // We have to handle the case where we have 0.0000somenumber.
-      const char *start = start_digits;
-      while ((*start == '0') || (*start == '.')) {
-        start++;
+    uint64_t total_digit_count = digit_count + exponent;
+    if (unlikely((total_digit_count > 19) ) ) { // this is uncommon
+      // might have an  overflow
+      int64_t deduct = 0;
+      if((digit_count == 1) && (*start_digits=='0')) {
+        deduct ++;
+        if(start_digits[1]=='.') {
+          const char * fap = start_digits + 2;
+          while(*fap == '0') {
+            fap++;
+          }
+          deduct += fap - (start_digits + 2);
+        }
       }
-      // we over-decrement by one when there is a '.'
-      digit_count -= (start - start_digits);
-      if (digit_count >= 19) {
-        // Ok, chances are good that we had an overflow!
-        // this is almost never going to get called!!!
-        // we start anew, going slowly!!!
+      total_digit_count -= deduct;
+      if(total_digit_count > 19) { // overflow
         return parse_float(buf, pj, offset, found_minus);
       }
     }
-    if (unlikely((power_index > 2 * 308))) { // this is uncommon!!!
-      // this is almost never going to get called!!!
-      // we start anew, going slowly!!!
-      return parse_float(buf, pj, offset, found_minus);
+    int64_t total_exponent = -exponent + exp_number;
+    double d = i; // as long as i is in [0, 2^53), this is exact
+    if (unlikely((total_exponent >  308) || (total_exponent <  -308))) {
+        return parse_float(buf, pj, offset, found_minus);
+    }
+    // as long as |total_exponent| < 22, this is exact
```

Comment:
We can cover the “hidden” precise cases as well, see the rapid JSON source code I linked in the issue and the original web article

Reply:
Right. When the total_exponent is too big, but our parsed integer mantissa is small, we can multiply the mantissa by a power of ten and then reduce total_exponent by that much.